src/components/street-segment.js

```diff
@@ -0,0 +1,424 @@
+          );
+        } else {
+          this.el.setAttribute(
+            `street-generated-stencil__${index}`,
+            `model: ${clone.model}; length: ${this.data.length}; spacing: ${clone.spacing}; direction: ${this.data.direction}; count: ${clone.count};`
+          );
+        }
+      });
+    }
+    if (componentsToGenerate?.pedestrians?.length > 0) {
+      componentsToGenerate.pedestrians.forEach((pedestrian, index) => {
+        this.el.setAttribute(
+          `street-generated-pedestrians__${index}`,
+          `segmentWidth: ${this.data.width}; density: ${pedestrian.density}; length: ${this.data.length}; direction: ${this.data.direction};`
+        );
+      });
+    }
+  },
+  updateSurfaceFromType: function (typeObject) {
+    // update color, surface, level from segment type preset
+    this.el.setAttribute(
+      'street-segment',
+      `surface: ${typeObject.surface}; color: ${typeObject.color}; level: ${typeObject.level};`
+    ); // to do: this should be more elegant to check for undefined and set default values
+  },
+  updateGeneratedComponentsList: function () {
+    // get all components on entity with prefix 'street-generated'
+    let generatedComponentList = [];
+    const components = this.el.components;
+    for (const componentName in components) {
+      if (componentName.startsWith('street-generated')) {
+        generatedComponentList.push(componentName);
+      }
+    }
+    this.generatedComponents = generatedComponentList;
+  },
+  update: function (oldData) {
+    const data = this.data;
+    const dataDiff = AFRAME.utils.diff(oldData, data);
+    // if oldData is same as current data, then don't update
+    if (AFRAME.utils.deepEqual(oldData, data)) {
+      return;
+    }
+    // regenerate components if only type has changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('type')
+    ) {
+      let typeObject = this.types[this.data.type];
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      this.remove();
+      this.createGeneratedComponentsFromType(typeObject); // add components for this type
+      this.updateSurfaceFromType(typeObject); // update surface color, surface, level
+    }
+    // propagate change of direction to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
+      Object.keys(dataDiff).includes('direction')
+    ) {
+      this.updateGeneratedComponentsList(); // if components were created through streetmix or streetplan import
+      for (const componentName of this.generatedComponents) {
+        this.el.setAttribute(componentName, 'direction', this.data.direction);
+      }
+    }
+    // propagate change of length to generated components is solo changed
+    if (
+      Object.keys(dataDiff).length === 1 &&
```

Comment:
Why the length === 1 check? We may update the street-segment component with two changed properties at the same time for example. It will probably not happen via the React UI though.

Reply:
b61593556baf55ea3731463fc84f798148446583